### DISCUSSION TITLE
Adding views, not only viewsets to Browsable API via routers

### DISCUSSION THREAD

**Discussion by BurkovBA** [encode/django-rest-framework#7830] [Ideas & Suggestions] [answered]
It is often desirable to be able to add views, not viewsets, to the Browsable API. For instance, see: http://stackoverflow.com/questions/18818179/routing-api-views-in-django-rest-framework.

There are custom router solutions, such as [this outdated HybridRouter implementation on stackoverflow](http://stackoverflow.com/questions/18817988/using-django-rest-frameworks-browsable-api-with-apiviews), but it would be nice to have this functionality out-of-the-box and consistent with evolution of DRF

**Comment by BurkovBA** [comment]
My current solution, based on [this post](http://stackoverflow.com/questions/18817988/using-django-rest-frameworks-browsable-api-with-apiviews), is as follows:

``` python
class HybridRouter(routers.DefaultRouter):
    def __init__(self, *args, **kwargs):
        super(HybridRouter, self).__init__(*args, **kwargs)
        self._api_view_urls = {}

    def add_api_view(self, name, url):
        self._api_view_urls[name] = url

    def remove_api_view(self, name):
        del self._api_view_urls[name]

    @property
    def api_view_urls(self):
        ret = {}
        ret.update(self._api_view_urls)
        return ret

    def get_urls(self):
        urls = super(HybridRouter, self).get_urls()
        for api_view_key in self._api_view_urls.keys():
            urls.append(self._api_view_urls[api_view_key])
        return urls

    def get_api_root_view(self):
        # Copy the following block from Default Router
        api_root_dict = {}
        list_name = self.routes[0].name
        for prefix, viewset, basename in self.registry:
            api_root_dict[prefix] = list_name.format(basename=basename)

        # In addition to that:
        api_view_urls = self._api_view_urls

        class APIRoot(views.APIView):
            _ignore_model_permissions = True

            def get(self, request, *args, **kwargs):
                ret = OrderedDict()
                namespace = request.resolver_match.namespace
                for key, url_name in api_root_dict.items():
                    if namespace:
                        url_name = namespace + ':' + url_name
                    try:
                        ret[key] = reverse(
                            url_name,
                            args=args,
                            kwargs=kwargs,
                            request=request,
                            format=kwargs.get('format', None)
                        )
                    except NoReverseMatch:
                        # Don't bail out if e.g. no list routes exist, only detail routes.
                        continue

                # In addition to what had been added, now add the APIView urls
                for api_view_key in api_view_urls.keys():
                    namespace = request.resolver_match.namespace
                    if namespace:
                        url_name = namespace + ":" + api_view_key
                    ret[api_view_key] = reverse(url_name,
                                        args=args,
                                        kwargs=kwargs,
                                        request=request,
                                        format=kwargs.get('format', None))

                return response.Response(ret)

        return APIRoot.as_view()
```

**Comment by lovelydinosaur** [comment]
Closing given lack of interest. We may reconsider if this is subsequently raised by other developers. Not sure I see the value over simply adding views explicitly to the URL conf.

**Comment by browze** [comment]
It seems to me that the value is that it would add those urls to the 'browsable' api?  Having just tried it and cast around on SO, it appears to me that something in the vein of the solution shown by BurkovBA above (or some other significant customisation) is still necessary in order to include certain 'single' views in the browsable API - for example a url that returns a single detail view and which doesn't require any kind of id to be passed (e.g. because its an aggregate view, or a singleton model).  Apologies if I'm missing something obvious. Thanks for DRF btw @tomchristie - it's great!

**Comment by xordoquy** [comment]
It feels quite complex to me while writing the same view as a `ViewSet` gets you the integration for free.
Don't forget that you can pick the desired behaviour for a `ViewSet` by implementing only a subset of the methods. For exemple, only implementing `list` or inheriting from `ListModelMixin` will not show details or creation for that `ViewSet`.

**Comment by BurkovBA** [comment]
@xordoquy True, but pure views (especially function-based) give you much more flexibility. E.g. you want to add api schema view to the browsable api, what do you do?

**Comment by browze** [comment]
On revisiting today, I quickly realised actually I don't have to customise to avoid passing an id if I set (or leave) detail as false on the @action decorator.  Last night I was a bit whacked and I guess I had wrongly assumed that if I didn't set detail = True it would only allow the result to be passed back as an array.  Thought I'd mention this for anyone else as dozy, happening on this thread.  

I don't know whether or not leaving detail = False is broadly inline with the 'idea' behind the flag when I'm presenting a view that isn't a collection of anything - however even it wasn't quite, that's probably quibbling...  For me this helps and would address many of my use cases even if occasionally I had to create an otherwise unnecessary viewset and I do get that it might keep the internals simpler..

However obviously there may be some cases (such as the api schema one just cited by @BurkovBA) where it might be necessary/really helpful to be able to install things that aren't viewset @action s at all and/or have them registered at root, - so just FWIW personally I still think it might be great to have that supported :)

**Comment by Reubentrapdoor** [comment]
I would like to add my token auth endpoint to the API browser so my users actually know where to find it, but since ObtainAuthToken inherits from APIView I can't find a good way to do that.

Am i supposed to just reimplement that whole thing as a viewset then just to be able to use DefaultRouter and get the URL to show up?

**Comment by rpkilby** [comment]
It's not an ideal solution, but you should be able to create a viewset that inherits `ObtainAuthToken`, then point `create` to `post` something like..

```python
class AuthTokenViewSet(ObtainAuthToken, ViewSet):
    def create(self, request, *args, **kwargs):
        # note that we need to call super here, so that we call `ObtainAuthToken.post`
        # instead of the `post` method set by `ViewSet.as_view()`
        return super().post(request, *args, **kwargs)
```

Let me know if that does/doesn't work.

**Comment by Reubentrapdoor** [comment]
That sees to work nicely for my use case thanks! Wouldnt have thought to use multiple inheritance myself.

**Comment by carltongibson** [comment, marked as answer]
Or, [subclass `DefaultRouter` and customise the `APIRootView`](https://github.com/encode/django-rest-framework/blob/37dcd553705421201f00fb7f6a75a6214afb9d60/rest_framework/routers.py#L341-L350) to explicitly add the auth view, [similar to how we manually define that in the tutorial](https://www.django-rest-framework.org/tutorial/5-relationships-and-hyperlinked-apis/#creating-an-endpoint-for-the-root-of-our-api).

**Comment by eldano1995** [comment]
I too like the idea of adding views to the browseable API using the router.

But this would be solved if there was a way to implement complex  path_url in the action inside a ModelViewSet. 

For example, let's say I have an UserViewSet and that I would like to have two additional endpoints: -

- one for requesting password reset (**users/request_password_reset**)
- the other for  actually resetting the password (**users/reset_password/<uidb64>/<token>**)

The fields uidb64 and token are used to confirm that the user clicked the reset_password_link that was sent to his/her email.

The previous is a common use case scenario. You could argue that given your are doing a POST request to **users/reset_password/<uidb64>/<token>/**, you could simplify the url to **users/reset_password/** and send uidb64 and token as POST params. 

But in a similar case, where you wouldn't use POST, like in **users/activate_account/<uidb64>/<token>/** (GET request), that wouldn't be possible.

That's why I think it would be great to allow to enable complex url_path in the action decorator function.

**Comment by auvipy** [comment]
> Closing given lack of interest. We may reconsider if this is subsequently raised by other developers. Not sure I see the value over simply adding views explicitly to the URL conf.

this could be moved to the discussion section

**Reply by lovelydinosaur** [reply]
Good call. Done.

**Comment by ghferrari** [comment]
I'd like to broadly add my support for this kind of use case. The overall issue seems to be that people want to add things to the api root view that is produced by `DefaultRouter` but they don't see an obvious way to do this using `ViewSets`. So I think there are two possibilities:

1. If there is an easy way to do this using the current implementation of `DefaultRouter` and `ViewSets`, then it would be useful to add this sort of case to the documentation.
2. If there is no easy way to do this using the current implementation of `DefaultRouter` and `ViewSets`, then it would great if `DefaultRouter` could be extended to allow registration of regular views rather than being restricted to `ViewSets`.

My own use case is that I want to be able to use `DefaultRouter` for lots of cases, but I may also have some API views that I want listed in the api root view but which aren't obviously handled by `ViewSets`.

It's true that I'm a newbie to DRF and so perhaps what I want can be easily handled using `ViewSets` - but if so, it would nice for the tutorial to have a section covering some examples. For example, I'd like to implement `/users/me` to return information about the currently logged in user. It seems I might be able to handle that with a custom action inside a `UserViewSet` and I'm planning to try it. But it would be good to see some examples of other problems being solved in this way.

The above solution might work in my case where I already want the default functionality of a  `UserViewSet` and then I'm happy to add an extra action to that `ViewSet`. But what about those cases where my api endpoint doesn't fit into a `ViewSet`? How can I add a regular `APIView` to the api root created by `DefaultRouter`? At the moment, when `DefaultRouter` only accepts `ViewSets`, that seems either impossible or to need a complex workaround like others have given above. It would be nice to have a simple, built-in way of doing this.

**Comment by andrelccorrea** [comment]
I would also like to be able to add Views and not just Viewsets to the Browseable Api.
From my point of view, Views are more explicit than ViewSets.

**Comment by goliney** [comment]
> We may reconsider if this is subsequently raised by other developers.

I came here to say that I would like to be able to add Views to the Browseable API too.

**Comment by kakulukia** [comment]
Being able to add APIViews to the browsable API would be awesome! Tried to do just that some moments ago and ended up here. :(
bummer

**Comment by radusuciu** [comment]
I would also use this feature on day one of release. My use case is a bit more complicated that most probably, but I use `drf-spectacular` and then `openapi-typescript-codegen` to generate types for use clientside. For most of my endpoints `ViewSet` and `DefaultRouter` work really nice and easy, but in cases where I have a view that isn't associated with a model (eg. I have a view that outputs statistics across a variety of models), there's no one workaround that allows me to keep both the ease of use of adding URLs with `DefaultRouter` and end-to-end type safety. If I implement a `ViewSet` with a `list` method, I have a hard time overriding the response type of the endpoint (it's not really a list!) without using `drf-spectacular` post-processing hooks, and if I implement a `retrieve` method instead, I have a different set of hacks I need to apply (eg. need to provide a dummy `queryset`, and then I can't override the request parameters..)

If I instead could write an `APIView`, and if it could be added to a router instance, I could have my cake and eat it too:
- browsable API integration
- modularized inclusion of URLs
- correct types

all with minimal hacks.

Edit: just wanted to add that of course I don't expect that DRF adds explicit support for my use-case, but I did want to highlight that while using `ViewSet` instead of `APIView` is simple enough, the semantics associated with a viewset can have downstream implications that make them not ideal.

**Comment by AStrayByte** [comment]
I ALSO want this. can please add this as it would do so much to increase the capability of DRF while not requiring too much work on the DRF team

**Comment by eatyourpeas** [comment]
I am happy to join this party too - my project has a series of custom endpoints that complement the Model Viewsets and would look nice in the browsable API grouped together.

**Comment by DevOpsCraftsman** [comment]
Supporting this, simplify sharing knowledge about endpoints within the teams without having to write superfluous documentation (that needs to be updated anytime something changes…)

**Comment by nomad-vagabond** [comment]
Me too. Would like to be able to add APIView url to the DRF browsable API in some simple way.

**Comment by Natgho** [comment]
I think this is a big shortcoming for the DRF. Defaultrouter and browsable API is a great and functional solution to make the project controlled by front-end developers. Please take action.

**Comment by andretheronsa** [comment]
Another vote from me. My use case - I would like to use DRF permissions / throttles / authentication / routing for an auth_check view that should just return status codes. I would like this to be in our browsable API & URL path (/api/v2/). I thought I would just be able to subclass APIView and get on it but apparently not. I think it's confusing for DRF to provide a view called "APIView" that cannot be added to the default API router. Most SO snippets for a hybrid router are not working anymore either - looks like `get_extra_actions` is the first isue. `AttributeError: type object 'AuthCheck' has no attribute 'get_extra_actions'`

**Comment by krugergui** [comment]
I'm in the exact use case as andre above, so another vote from me.

**Comment by hastaka** [comment]
Found a working solution to put non-viewset views on the root page of a DefaultRouter, based on carltongibson's suggestion above:
(Python 3.11.5, Django 4.2.5, djangorestframework 3.14.0)

In urls.py:
```
# Create a router subclassed from DefaultRouter and give it a list[dict] as input
class APIRouter(routers.DefaultRouter):
    '''
    Customized Default Router to include non-viewset views on root page
    '''
    singleViews:list
    def __init__(self, singleViews:list, *args, **kwargs):
        self.singleViews = singleViews
        super().__init__(*args, **kwargs)

    def get_api_root_view(self, api_urls=None):
        """
        Return a basic root view.
        """
        api_root_dict = OrderedDict()
        list_name = self.routes[0].name
        for prefix, viewset, basename in self.registry:
            api_root_dict[prefix] = list_name.format(basename=basename)
        for singleView in self.singleViews:
            api_root_dict[singleView['route']] = singleView['name']
        return self.APIRootView.as_view(api_root_dict=api_root_dict)

# Define list - replace my_api_view with any view class that subclasses DRF's APIView (including Login, Logout views e.g. django-rest-knox)
singleViews = [
    {
        'route': 'login',
        'view': views.LoginView.as_view(),
        'name': 'knox_login'
    },
    {
        'route': 'logout',
        'view': knox_views.LogoutView.as_view(),
        'name': 'knox_logout'
    },
    {
        'route': 'logoutall',
        'view': knox_views.LogoutAllView.as_view(),
        'name': 'knox_logoutall'
    },
    {
        'route': 'my-api-view-route',
        'view': views.MyAPIView.as_view(),
        'name': 'my-api-view'
    }, # etc...
]

# Normal router setup, including list and registering ViewSets
router = APIRouter(singleViews=singleViews)
router.register('my-viewset', views.MyViewSet, basename='myviewset') # etc...

# Set up URL Patterns as normal (with router include + any URLs not included on root page) and extend with single views to register
urlpatterns = [
    path('', include(router.urls)),
    path('api-auth/', include('rest_framework.urls', namespace='rest_framework')),
    path('admin/', admin.site.urls),
]
urlpatterns.extend([path(route=i['route'], view=i['view'], name=i['name']) for i in singleViews])
```

**Reply by krugergui** [reply]
Works great! Thanks!
